frontend/modules/compare/compare.sagas.ts

```diff
@@ -0,0 +1,414 @@
+import { selectRevisions, selectIsFederation, selectSettings, ModelTypes } from '../model';
+import { COMPARE_SORT_TYPES, DIFF_COMPARE_TYPE, RENDERING_TYPES, VULNERABLE_PROPS } from '../../constants/compare';
+import { DialogActions } from '../dialog';
+import { Viewer } from '../../services/viewer/viewer';
+import { SORT_ORDER_TYPES } from '../../constants/sorting';
+import * as API from '../../services/api';
+import { getAngularService } from '../../helpers/migration';
+import {
+	selectSortType,
+	selectSortOrder,
+	selectIsCompareActive,
+	selectActiveTab,
+	selectSelectedModelsMap,
+	selectBaseModelsList,
+	selectTargetClashModels,
+	selectTargetDiffModels,
+	selectTargetModelsList,
+	selectComponentState,
+	selectCompareModels
+} from './compare.selectors';
+
+const getNextRevision = (revisions, currentRevision) => {
+	if (!currentRevision) {
+		return revisions[0];
+	}
+
+	const index = revisions.findIndex((r) => r._id === currentRevision._id);
+	const lastRev = index + 1 === revisions.length;
+
+	if (lastRev) {
+		return revisions[index];
+	}
+
+	return revisions[index + 1];
+};
+
+const prepareModelToCompare = (teamspace, modelId, name, isFederation, revisions, currentRevision?) => {
+	const baseRevision = isFederation
+		? revisions[0]
+		: revisions.find((rev) => rev.tag === currentRevision || rev._id === currentRevision ) || revisions[0];
+
+	const targetRevision = getNextRevision(revisions, baseRevision);
+	const revisionData = currentRevision
+		? revisions.find((revision) => revision.tag === currentRevision)
+		: revisions[0];
+
+	return {
+		_id: modelId,
+		teamspace,
+		name,
+		baseRevision,
+		currentRevision: revisionData,
+		targetDiffRevision: targetRevision,
+		targetClashRevision: baseRevision,
+		revisions
+	};
+};
+
+function* getCompareModelData({ isFederation, settings }) {
+	try {
+		const revisions = yield select(selectRevisions);
+		const [teamspace, , currentRevision] = window.location.pathname.replace('/viewer/', '').split('/');
+
+		if (!isFederation) {
+			const model =
+				prepareModelToCompare(teamspace, settings._id, settings.name, isFederation, revisions, currentRevision);
+			yield put(CompareActions.setComponentState({ compareModels: [model], isPending: false }));
+		} else {
+			const { data: submodelsRevisionsMap } = yield API.getSubModelsRevisions(teamspace, settings._id, currentRevision);
+			const compareModels = settings.subModels.map(({ model }) => {
+				const subModelData = submodelsRevisionsMap[model];
+				return prepareModelToCompare(teamspace, model, subModelData.name, false, subModelData.revisions);
+			});
+
+			const selectedModelsMap = compareModels.reduce((map, model) => {
+				map[model._id] = true;
+				return map;
+			}, {});
+
+			yield put(CompareActions.setComponentState({
+				compareModels,
+				selectedDiffModelsMap: selectedModelsMap,
+				selectedClashModelsMap: selectedModelsMap,
+				targetDiffModels: selectedModelsMap,
```

Comment:
Because this is just setting the initial state of our _Compare_ store data.

Reply:
What @sanmont3drepo was wondering is that why do we need 3 references to begin with. Unless we have misunderstood, it seems unnecessary to have all 3 if they are just references to the same object? It would make sense to separate them if you are making a copy of the objects. If I make changes to one, i am effectively making changes to all, so doing [things like this](https://github.com/3drepo/3drepo.io/pull/1467/files#diff-7b9b81947db4f121e3a64bfe6454017fR69) seems pointless.

I feel like we must be missing something! 😆

Reply:
It didn't matter and works good because it was just initialState and called once on start, but you are right, we can use here the spread operator to make sure that everything will be clean :)